src/docfx/lib/HtmlUtility.cs

```diff
@@ -163,6 +164,61 @@ public static string TransformLinks(string html, Func<string, string> transform)
             return result.ToString();
         }
 
+        public static (List<(string uid, int line, int column, string errorType)>, string) TransformXref(string html, int lineNumber, Func<string, (Error error, string href, string display, Document file)> transform)
+        {
+            var errors = new List<(string uid, int line, int column, string errorType)>();
+
+            // Fast pass it does not have <xref> tag
+            if (!(html.Contains("<xref", StringComparison.OrdinalIgnoreCase) && html.Contains("href", StringComparison.OrdinalIgnoreCase)))
+            {
+                return (errors, html);
+            }
+
+            var doc = new HtmlDocument();
+            doc.LoadHtml(html);
+
+            // TODO: get accurate line and column for HTML block lasting several lines and multiple nodes in the same line
+            var replacingNodes = new List<(HtmlNode, HtmlNode)>();
+            foreach (var node in doc.DocumentNode.Descendants())
+            {
+                if (node.Name != "xref")
+                {
+                    continue;
+                }
+
+                var xref = node.Attributes["href"];
+                if (xref is null)
+                {
+                    continue;
+                }
+
+                // data-throw-if-not-resolved from v2 is not needed any more since we can decide if warning throw by checking raw
+                var rawSource = node.GetAttributeValue("data-raw-source", null);
+                var rawHtml = node.GetAttributeValue("data-raw-html", null);
+                var raw = HttpUtility.HtmlDecode(!string.IsNullOrEmpty(rawHtml) ? rawHtml : rawSource);
+                var (_, resolvedHref, display, _) = transform(HttpUtility.HtmlDecode(xref.Value));
+
+                var resolvedNode = new HtmlDocument();
+                if (string.IsNullOrEmpty(resolvedHref))
+                {
+                    errors.Add((xref.Value, lineNumber, s_getValueStartIndex(xref), raw.StartsWith("@") ? nameof(Errors.AtUidNotFound) : nameof(Errors.UidNotFound)));
+                    resolvedNode.LoadHtml(raw);
+                }
+                else
+                {
+                    resolvedNode.LoadHtml($"<a href='{HttpUtility.HtmlEncode(resolvedHref)}'>{display}</a>");
```

Comment:
do you also need to encode `{display}` in case it contains `<`? Also add a test case about encoding handling in `encoding.yml`?

Reply:
Actually, I am planning to do this within `Xrefmap.Resolve()`, decode `href` and encode `resolveXref` and `display`, feel it should be in a later PR